pkg/cloud/externalconn/connection_record.go

```diff
@@ -0,0 +1,278 @@
+func (e *ExternalConnection) InitFromDatums(
+	datums []tree.Datum, cols []colinfo.ResultColumn,
+) error {
+	if len(datums) != len(cols) {
+		return errors.Errorf(
+			"datums length != columns length: %d != %d", len(datums), len(cols))
+	}
+
+	record := reflect.ValueOf(&e.rec).Elem()
+
+	numInitialized := 0
+	for i, col := range cols {
+		native, err := datumToNative(datums[i])
+		if err != nil {
+			return err
+		}
+
+		if native == nil {
+			continue
+		}
+
+		fieldNum, ok := columnNameToField[col.Name]
+		if !ok {
+			// Table contains columns we don't care about (e.g. created)
+			continue
+		}
+
+		field := record.Field(fieldNum)
+
+		if data, ok := native.([]byte); ok {
+			// []byte == protocol message.
+			if pb, ok := field.Addr().Interface().(protoutil.Message); ok {
+				if err := protoutil.Unmarshal(data, pb); err != nil {
+					return err
+				}
+			} else {
+				return errors.Newf(
+					"field %s with value of type %T is does not appear to be a protocol message",
+					field.String(), field.Addr().Interface())
+			}
+		} else {
+			// We ought to be able to assign native directly to our field.
+			// But, be paranoid and double check.
+			rv := reflect.ValueOf(native)
+			if !rv.Type().AssignableTo(field.Type()) {
+				return errors.Newf("value of type %T cannot be assigned to %s",
+					native, field.Type().String())
+			}
+			field.Set(rv)
+		}
+		numInitialized++
+	}
+
+	if numInitialized == 0 {
+		return errors.New("did not initialize any external connection field")
+	}
+
+	return nil
+}
+
+// generates "$1,$2,..." placeholders for the specified 'n' number of arguments.
+func generatePlaceholders(n int) string {
+	placeholders := strings.Builder{}
+	for i := 1; i <= n; i++ {
+		if i > 1 {
+			placeholders.WriteByte(',')
+		}
+		placeholders.WriteString(fmt.Sprintf("$%d", i))
+	}
+	return placeholders.String()
+}
+
+// Create persists this external connection in the system.external_connections
+// table.
+//
+// Only the values initialized in the receiver are persisted in the system
+// table. If an error is returned, it is callers responsibility to handle it
+// (e.g. rollback transaction).
+func (e *ExternalConnection) Create(
+	ctx context.Context, ex sqlutil.InternalExecutor, txn *kv.Txn,
+) error {
+	cols, qargs, err := e.marshalChanges()
+	if err != nil {
+		return err
+	}
+
+	createQuery := "INSERT INTO system.external_connections (%s) VALUES (%s) RETURNING connection_name"
+	row, retCols, err := ex.QueryRowExWithCols(ctx, "ExternalConnection.Create", txn,
+		sessiondata.InternalExecutorOverride{User: username.RootUserName()},
+		fmt.Sprintf(createQuery, strings.Join(cols, ","), generatePlaceholders(len(qargs))),
+		qargs...,
+	)
+	if err != nil {
+		// If we see a UniqueViolation it means that we are attempting to create an
+		// External Connection with a `connection_name` that already exists.
+		if pgerror.GetPGCode(err) == pgcode.UniqueViolation {
```

Comment:
I think the main concern was around the renaming of External Connections. However, @dt rightly pointed out that we should not allow this since it opens us up to having to rewrite all persisted references of this object. 

As far as reusing a name goes: this should be tackled by the v2 work around not allowing a drop if someone is actively using the connection.